nixos/modules/services/networking/firewalld/zone.nix

```diff
@@ -0,0 +1,275 @@
+            };
+          });
+          description = "Source addresses, address ranges, MAC addresses or ipsets to bind.";
+          default = [ ];
+        };
+        icmpBlockInversion = mkOption {
+          type = bool;
+          description = ''
+            Whether to invert the icmp block handling.  Only enabled ICMP types
+            are accepted and all others are rejected in the zone.
+          '';
+          default = false;
+        };
+        forward = mkOption {
+          type = bool;
+          description = ''
+            Whether to enable intra-zone forwarding.  When enabled, packets
+            will be forwarded between interfaces or sources within a zone, even
+            if the zone's target is not set to ACCEPT.
+          '';
+          default = false;
+        };
+        short = mkOption {
+          type = nullOr nonEmptyStr;
+          description = "Short description for the zone.";
+          default = null;
+        };
+        description = mkOption {
+          type = nullOr nonEmptyStr;
+          description = "Description for the zone.";
+          default = null;
+        };
+        services = mkOption {
+          type = listOf nonEmptyStr;
+          description = "Services to allow in the zone.";
+          default = [ ];
+        };
+        ports = mkOption {
+          type = listOf (submodule portProtocolOptions);
+          description = "Ports to allow in the zone.";
+          default = [ ];
+        };
+        protocols = mkOption {
+          type = listOf nonEmptyStr;
+          description = "Protocols to allow in the zone.";
+          default = [ ];
+        };
+        icmpBlocks = mkOption {
+          type = listOf nonEmptyStr;
+          description = "ICMP types to block in the zone.";
+          default = [ ];
+        };
+        masquerade = mkOption {
+          type = bool;
+          description = "Whether to enable masquerading in the zone.";
+          default = false;
+        };
+        forwardPorts = mkOption {
+          type = listOf (submodule {
+            options = {
+              port = mkPortOption { };
+              protocol = protocolOption;
+              to-port = (mkPortOption { optional = true; }) // {
+                default = null;
+              };
+              to-addr = mkOption {
+                type = nullOr nonEmptyStr;
+                description = "Destination IP address.";
+                default = null;
+              };
+            };
+          });
+          description = "Ports to forward in the zone.";
+          default = [ ];
+        };
+        sourcePorts = mkOption {
+          type = listOf (submodule portProtocolOptions);
+          description = "Source ports to allow in the zone.";
+          default = [ ];
+        };
+        rules = mkOption {
+          type = listOf (format.type);
+          description = "Rich rules for the zone.";
+          default = [ ];
+        };
+      };
+    });
```

Comment:
I wonder if it'd be possible to re-think a lot of these large option declarations with an RFC42-style freeform option?

I've not looked closely enough at the specific lines I'm commenting on. Rather, I'm asking more generally about the firewalld options more generally, whether some of them could be less explicit and more freeform?

Freeform options are great for increasing flexibility and reducing maintenance burden.

Reply:
The problem here is that you need tricky conversions to make a valid XML.  I would be happy if there's another way but I haven't found one yet.  On the other hand the types are loosely based on the manual pages so I don't think that would be a problem.